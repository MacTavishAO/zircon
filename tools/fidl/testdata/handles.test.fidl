library fidl.test.handles;

// This is not a reference to library zx to avoid complexity in the test data.
// A local handle is defined here that contains subtypes that this test cares
// about.
enum obj_type : uint32 {
    NONE = 0;
    PROCESS = 1;
    THREAD = 2;
    VMO = 3;
    CHANNEL = 4;
    EVENT = 5;
    PORT = 6;
    INTERRUPT = 9;
    PCI_DEVICE = 11;
    LOG = 12;
    SOCKET = 14;
    RESOURCE = 15;
    EVENTPAIR = 16;
    JOB = 17;
    VMAR = 18;
    FIFO = 19;
    GUEST = 20;
    VCPU = 21;
    TIMER = 22;
    IOMMU = 23;
    BTI = 24;
    PROFILE = 25;
    PMT = 26;
    SUSPEND_TOKEN = 27;
    PAGER = 28;
    EXCEPTION = 29;
    CLOCK = 30;
    STREAM = 31;
    MSI_ALLOCATION = 32;
    MSI_INTERRUPT = 33;
};

resource handle : uint32 {
    properties {
        obj_type subtype;
    };
};

protocol SomeProtocol {
};

using aliased_plain_handle = handle;
using aliased_subtype_handle = handle:VMO;
using aliased_rights_handle = handle<vmo, 1>;

struct Handles {
    handle plain_handle;

    handle:BTI bti_handle;
    handle:CHANNEL channel_handle;
    handle:CLOCK clock_handle;
    handle:LOG debuglog_handle;
    handle:EVENT event_handle;
    handle:EVENTPAIR eventpair_handle;
    handle:EXCEPTION exception_handle;
    handle:FIFO fifo_handle;
    handle:GUEST guest_handle;
    handle:INTERRUPT interrupt_handle;
    handle:IOMMU iommu_handle;
    handle:JOB job_handle;
    handle:PAGER pager_handle;
    handle:PCI_DEVICE pcidevice_handle;
    handle:PMT pmt_handle;
    handle:PORT port_handle;
    handle:PROCESS process_handle;
    handle:PROFILE profile_handle;
    handle:RESOURCE resource_handle;
    handle:SOCKET socket_handle;
    handle:SUSPEND_TOKEN suspendtoken_handle;
    handle:THREAD thread_handle;
    handle:TIMER timer_handle;
    handle:VCPU vcpu_handle;
    handle:VMAR vmar_handle;
    handle:VMO vmo_handle;

    handle<vmo, 1> rights_handle;

    aliased_plain_handle aliased_plain_handle_field;
    aliased_subtype_handle aliased_subtype_handle_field;
    aliased_rights_handle aliased_rights_handle_field;

    SomeProtocol some_protocol;
    request<SomeProtocol> request_some_protocol;
};
