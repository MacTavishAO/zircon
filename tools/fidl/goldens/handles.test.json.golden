{
  "version": "0.0.1",
  "name": "fidl.test.handles",
  "library_dependencies": [],
  "bits_declarations": [],
  "const_declarations": [],
  "enum_declarations": [
    {
      "name": "fidl.test.handles/obj_type",
      "location": {
        "filename": "handles.test.fidl",
        "line": 6,
        "column": 6,
        "length": 8
      },
      "type": "uint32",
      "members": [
        {
          "name": "NONE",
          "location": {
            "filename": "handles.test.fidl",
            "line": 7,
            "column": 5,
            "length": 4
          },
          "value": {
            "kind": "literal",
            "value": "0",
            "expression": "0",
            "literal": {
              "kind": "numeric",
              "value": "0",
              "expression": "0"
            }
          }
        },
        {
          "name": "PROCESS",
          "location": {
            "filename": "handles.test.fidl",
            "line": 8,
            "column": 5,
            "length": 7
          },
          "value": {
            "kind": "literal",
            "value": "1",
            "expression": "1",
            "literal": {
              "kind": "numeric",
              "value": "1",
              "expression": "1"
            }
          }
        },
        {
          "name": "THREAD",
          "location": {
            "filename": "handles.test.fidl",
            "line": 9,
            "column": 5,
            "length": 6
          },
          "value": {
            "kind": "literal",
            "value": "2",
            "expression": "2",
            "literal": {
              "kind": "numeric",
              "value": "2",
              "expression": "2"
            }
          }
        },
        {
          "name": "VMO",
          "location": {
            "filename": "handles.test.fidl",
            "line": 10,
            "column": 5,
            "length": 3
          },
          "value": {
            "kind": "literal",
            "value": "3",
            "expression": "3",
            "literal": {
              "kind": "numeric",
              "value": "3",
              "expression": "3"
            }
          }
        },
        {
          "name": "CHANNEL",
          "location": {
            "filename": "handles.test.fidl",
            "line": 11,
            "column": 5,
            "length": 7
          },
          "value": {
            "kind": "literal",
            "value": "4",
            "expression": "4",
            "literal": {
              "kind": "numeric",
              "value": "4",
              "expression": "4"
            }
          }
        },
        {
          "name": "EVENT",
          "location": {
            "filename": "handles.test.fidl",
            "line": 12,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "5",
            "expression": "5",
            "literal": {
              "kind": "numeric",
              "value": "5",
              "expression": "5"
            }
          }
        },
        {
          "name": "PORT",
          "location": {
            "filename": "handles.test.fidl",
            "line": 13,
            "column": 5,
            "length": 4
          },
          "value": {
            "kind": "literal",
            "value": "6",
            "expression": "6",
            "literal": {
              "kind": "numeric",
              "value": "6",
              "expression": "6"
            }
          }
        },
        {
          "name": "INTERRUPT",
          "location": {
            "filename": "handles.test.fidl",
            "line": 14,
            "column": 5,
            "length": 9
          },
          "value": {
            "kind": "literal",
            "value": "9",
            "expression": "9",
            "literal": {
              "kind": "numeric",
              "value": "9",
              "expression": "9"
            }
          }
        },
        {
          "name": "PCI_DEVICE",
          "location": {
            "filename": "handles.test.fidl",
            "line": 15,
            "column": 5,
            "length": 10
          },
          "value": {
            "kind": "literal",
            "value": "11",
            "expression": "11",
            "literal": {
              "kind": "numeric",
              "value": "11",
              "expression": "11"
            }
          }
        },
        {
          "name": "LOG",
          "location": {
            "filename": "handles.test.fidl",
            "line": 16,
            "column": 5,
            "length": 3
          },
          "value": {
            "kind": "literal",
            "value": "12",
            "expression": "12",
            "literal": {
              "kind": "numeric",
              "value": "12",
              "expression": "12"
            }
          }
        },
        {
          "name": "SOCKET",
          "location": {
            "filename": "handles.test.fidl",
            "line": 17,
            "column": 5,
            "length": 6
          },
          "value": {
            "kind": "literal",
            "value": "14",
            "expression": "14",
            "literal": {
              "kind": "numeric",
              "value": "14",
              "expression": "14"
            }
          }
        },
        {
          "name": "RESOURCE",
          "location": {
            "filename": "handles.test.fidl",
            "line": 18,
            "column": 5,
            "length": 8
          },
          "value": {
            "kind": "literal",
            "value": "15",
            "expression": "15",
            "literal": {
              "kind": "numeric",
              "value": "15",
              "expression": "15"
            }
          }
        },
        {
          "name": "EVENTPAIR",
          "location": {
            "filename": "handles.test.fidl",
            "line": 19,
            "column": 5,
            "length": 9
          },
          "value": {
            "kind": "literal",
            "value": "16",
            "expression": "16",
            "literal": {
              "kind": "numeric",
              "value": "16",
              "expression": "16"
            }
          }
        },
        {
          "name": "JOB",
          "location": {
            "filename": "handles.test.fidl",
            "line": 20,
            "column": 5,
            "length": 3
          },
          "value": {
            "kind": "literal",
            "value": "17",
            "expression": "17",
            "literal": {
              "kind": "numeric",
              "value": "17",
              "expression": "17"
            }
          }
        },
        {
          "name": "VMAR",
          "location": {
            "filename": "handles.test.fidl",
            "line": 21,
            "column": 5,
            "length": 4
          },
          "value": {
            "kind": "literal",
            "value": "18",
            "expression": "18",
            "literal": {
              "kind": "numeric",
              "value": "18",
              "expression": "18"
            }
          }
        },
        {
          "name": "FIFO",
          "location": {
            "filename": "handles.test.fidl",
            "line": 22,
            "column": 5,
            "length": 4
          },
          "value": {
            "kind": "literal",
            "value": "19",
            "expression": "19",
            "literal": {
              "kind": "numeric",
              "value": "19",
              "expression": "19"
            }
          }
        },
        {
          "name": "GUEST",
          "location": {
            "filename": "handles.test.fidl",
            "line": 23,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "20",
            "expression": "20",
            "literal": {
              "kind": "numeric",
              "value": "20",
              "expression": "20"
            }
          }
        },
        {
          "name": "VCPU",
          "location": {
            "filename": "handles.test.fidl",
            "line": 24,
            "column": 5,
            "length": 4
          },
          "value": {
            "kind": "literal",
            "value": "21",
            "expression": "21",
            "literal": {
              "kind": "numeric",
              "value": "21",
              "expression": "21"
            }
          }
        },
        {
          "name": "TIMER",
          "location": {
            "filename": "handles.test.fidl",
            "line": 25,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "22",
            "expression": "22",
            "literal": {
              "kind": "numeric",
              "value": "22",
              "expression": "22"
            }
          }
        },
        {
          "name": "IOMMU",
          "location": {
            "filename": "handles.test.fidl",
            "line": 26,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "23",
            "expression": "23",
            "literal": {
              "kind": "numeric",
              "value": "23",
              "expression": "23"
            }
          }
        },
        {
          "name": "BTI",
          "location": {
            "filename": "handles.test.fidl",
            "line": 27,
            "column": 5,
            "length": 3
          },
          "value": {
            "kind": "literal",
            "value": "24",
            "expression": "24",
            "literal": {
              "kind": "numeric",
              "value": "24",
              "expression": "24"
            }
          }
        },
        {
          "name": "PROFILE",
          "location": {
            "filename": "handles.test.fidl",
            "line": 28,
            "column": 5,
            "length": 7
          },
          "value": {
            "kind": "literal",
            "value": "25",
            "expression": "25",
            "literal": {
              "kind": "numeric",
              "value": "25",
              "expression": "25"
            }
          }
        },
        {
          "name": "PMT",
          "location": {
            "filename": "handles.test.fidl",
            "line": 29,
            "column": 5,
            "length": 3
          },
          "value": {
            "kind": "literal",
            "value": "26",
            "expression": "26",
            "literal": {
              "kind": "numeric",
              "value": "26",
              "expression": "26"
            }
          }
        },
        {
          "name": "SUSPEND_TOKEN",
          "location": {
            "filename": "handles.test.fidl",
            "line": 30,
            "column": 5,
            "length": 13
          },
          "value": {
            "kind": "literal",
            "value": "27",
            "expression": "27",
            "literal": {
              "kind": "numeric",
              "value": "27",
              "expression": "27"
            }
          }
        },
        {
          "name": "PAGER",
          "location": {
            "filename": "handles.test.fidl",
            "line": 31,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "28",
            "expression": "28",
            "literal": {
              "kind": "numeric",
              "value": "28",
              "expression": "28"
            }
          }
        },
        {
          "name": "EXCEPTION",
          "location": {
            "filename": "handles.test.fidl",
            "line": 32,
            "column": 5,
            "length": 9
          },
          "value": {
            "kind": "literal",
            "value": "29",
            "expression": "29",
            "literal": {
              "kind": "numeric",
              "value": "29",
              "expression": "29"
            }
          }
        },
        {
          "name": "CLOCK",
          "location": {
            "filename": "handles.test.fidl",
            "line": 33,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "30",
            "expression": "30",
            "literal": {
              "kind": "numeric",
              "value": "30",
              "expression": "30"
            }
          }
        },
        {
          "name": "STREAM",
          "location": {
            "filename": "handles.test.fidl",
            "line": 34,
            "column": 5,
            "length": 6
          },
          "value": {
            "kind": "literal",
            "value": "31",
            "expression": "31",
            "literal": {
              "kind": "numeric",
              "value": "31",
              "expression": "31"
            }
          }
        },
        {
          "name": "MSI_ALLOCATION",
          "location": {
            "filename": "handles.test.fidl",
            "line": 35,
            "column": 5,
            "length": 14
          },
          "value": {
            "kind": "literal",
            "value": "32",
            "expression": "32",
            "literal": {
              "kind": "numeric",
              "value": "32",
              "expression": "32"
            }
          }
        },
        {
          "name": "MSI_INTERRUPT",
          "location": {
            "filename": "handles.test.fidl",
            "line": 36,
            "column": 5,
            "length": 13
          },
          "value": {
            "kind": "literal",
            "value": "33",
            "expression": "33",
            "literal": {
              "kind": "numeric",
              "value": "33",
              "expression": "33"
            }
          }
        }
      ],
      "strict": true
    }
  ],
  "experimental_resource_declarations": [
    {
      "name": "fidl.test.handles/handle",
      "location": {
        "filename": "handles.test.fidl",
        "line": 39,
        "column": 10,
        "length": 6
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint32"
      },
      "properties": [
        {
          "name": "subtype",
          "location": {
            "filename": "handles.test.fidl",
            "line": 41,
            "column": 18,
            "length": 7
          },
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.handles/obj_type",
            "nullable": false
          }
        }
      ]
    }
  ],
  "interface_declarations": [
    {
      "name": "fidl.test.handles/SomeProtocol",
      "location": {
        "filename": "handles.test.fidl",
        "line": 45,
        "column": 10,
        "length": 12
      },
      "methods": []
    }
  ],
  "service_declarations": [],
  "struct_declarations": [
    {
      "name": "fidl.test.handles/Handles",
      "location": {
        "filename": "handles.test.fidl",
        "line": 52,
        "column": 8,
        "length": 7
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "handle",
            "subtype": "handle",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "plain_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 53,
            "column": 12,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "bti",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "bti_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 55,
            "column": 16,
            "length": 10
          },
          "field_shape_v1": {
            "offset": 4,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "channel",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "channel_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 56,
            "column": 20,
            "length": 14
          },
          "field_shape_v1": {
            "offset": 8,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "clock",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "clock_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 57,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 12,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "debuglog",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "debuglog_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 58,
            "column": 16,
            "length": 15
          },
          "field_shape_v1": {
            "offset": 16,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "event",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "event_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 59,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 20,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "eventpair",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "eventpair_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 60,
            "column": 22,
            "length": 16
          },
          "field_shape_v1": {
            "offset": 24,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "exception",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "exception_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 61,
            "column": 22,
            "length": 16
          },
          "field_shape_v1": {
            "offset": 28,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "fifo",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "fifo_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 62,
            "column": 17,
            "length": 11
          },
          "field_shape_v1": {
            "offset": 32,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "guest",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "guest_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 63,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 36,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "interrupt",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "interrupt_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 64,
            "column": 22,
            "length": 16
          },
          "field_shape_v1": {
            "offset": 40,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "iommu",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "iommu_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 65,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 44,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "job",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "job_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 66,
            "column": 16,
            "length": 10
          },
          "field_shape_v1": {
            "offset": 48,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "pager",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "pager_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 67,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 52,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "pcidevice",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "pcidevice_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 68,
            "column": 23,
            "length": 16
          },
          "field_shape_v1": {
            "offset": 56,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "pmt",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "pmt_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 69,
            "column": 16,
            "length": 10
          },
          "field_shape_v1": {
            "offset": 60,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "port",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "port_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 70,
            "column": 17,
            "length": 11
          },
          "field_shape_v1": {
            "offset": 64,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "process",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "process_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 71,
            "column": 20,
            "length": 14
          },
          "field_shape_v1": {
            "offset": 68,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "profile",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "profile_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 72,
            "column": 20,
            "length": 14
          },
          "field_shape_v1": {
            "offset": 72,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "resource",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "resource_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 73,
            "column": 21,
            "length": 15
          },
          "field_shape_v1": {
            "offset": 76,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "socket",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "socket_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 74,
            "column": 19,
            "length": 13
          },
          "field_shape_v1": {
            "offset": 80,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "suspendtoken",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "suspendtoken_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 75,
            "column": 26,
            "length": 19
          },
          "field_shape_v1": {
            "offset": 84,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "thread",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "thread_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 76,
            "column": 19,
            "length": 13
          },
          "field_shape_v1": {
            "offset": 88,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "timer",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "timer_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 77,
            "column": 18,
            "length": 12
          },
          "field_shape_v1": {
            "offset": 92,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "vcpu",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "vcpu_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 78,
            "column": 17,
            "length": 11
          },
          "field_shape_v1": {
            "offset": 96,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "vmar",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "vmar_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 79,
            "column": 17,
            "length": 11
          },
          "field_shape_v1": {
            "offset": 100,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "vmo",
            "rights": 2147483648,
            "nullable": false
          },
          "name": "vmo_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 80,
            "column": 16,
            "length": 10
          },
          "field_shape_v1": {
            "offset": 104,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "vmo",
            "rights": 1,
            "nullable": false
          },
          "name": "rights_handle",
          "location": {
            "filename": "handles.test.fidl",
            "line": 82,
            "column": 20,
            "length": 13
          },
          "field_shape_v1": {
            "offset": 108,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "handle",
            "rights": 2147483648,
            "nullable": false
          },
          "experimental_maybe_from_type_alias": {
            "name": "fidl.test.handles/aliased_plain_handle",
            "args": [],
            "nullable": false
          },
          "name": "aliased_plain_handle_field",
          "location": {
            "filename": "handles.test.fidl",
            "line": 84,
            "column": 26,
            "length": 26
          },
          "field_shape_v1": {
            "offset": 112,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "handle",
            "rights": 2147483648,
            "nullable": false
          },
          "experimental_maybe_from_type_alias": {
            "name": "fidl.test.handles/aliased_subtype_handle",
            "args": [],
            "nullable": false
          },
          "name": "aliased_subtype_handle_field",
          "location": {
            "filename": "handles.test.fidl",
            "line": 85,
            "column": 28,
            "length": 28
          },
          "field_shape_v1": {
            "offset": 116,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "handle",
            "subtype": "handle",
            "rights": 1,
            "nullable": false
          },
          "experimental_maybe_from_type_alias": {
            "name": "fidl.test.handles/aliased_rights_handle",
            "args": [],
            "nullable": false
          },
          "name": "aliased_rights_handle_field",
          "location": {
            "filename": "handles.test.fidl",
            "line": 86,
            "column": 27,
            "length": 27
          },
          "field_shape_v1": {
            "offset": 120,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.handles/SomeProtocol",
            "nullable": false
          },
          "name": "some_protocol",
          "location": {
            "filename": "handles.test.fidl",
            "line": 88,
            "column": 18,
            "length": 13
          },
          "field_shape_v1": {
            "offset": 124,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "request",
            "subtype": "fidl.test.handles/SomeProtocol",
            "nullable": false
          },
          "name": "request_some_protocol",
          "location": {
            "filename": "handles.test.fidl",
            "line": 89,
            "column": 27,
            "length": 21
          },
          "field_shape_v1": {
            "offset": 128,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 132,
        "alignment": 4,
        "depth": 0,
        "max_handles": 33,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": true
      }
    }
  ],
  "table_declarations": [],
  "union_declarations": [],
  "type_alias_declarations": [
    {
      "name": "fidl.test.handles/aliased_plain_handle",
      "location": {
        "filename": "handles.test.fidl",
        "line": 48,
        "column": 7,
        "length": 20
      },
      "partial_type_ctor": {
        "name": "handle",
        "args": [],
        "nullable": false
      }
    },
    {
      "name": "fidl.test.handles/aliased_subtype_handle",
      "location": {
        "filename": "handles.test.fidl",
        "line": 49,
        "column": 7,
        "length": 22
      },
      "partial_type_ctor": {
        "name": "handle",
        "args": [],
        "nullable": false
      }
    },
    {
      "name": "fidl.test.handles/aliased_rights_handle",
      "location": {
        "filename": "handles.test.fidl",
        "line": 50,
        "column": 7,
        "length": 21
      },
      "partial_type_ctor": {
        "name": "handle",
        "args": [],
        "nullable": false,
        "maybe_handle_subtype": "vmo",
        "handle_rights": {
          "kind": "literal",
          "value": "1",
          "expression": "1",
          "literal": {
            "kind": "numeric",
            "value": "1",
            "expression": "1"
          }
        }
      }
    }
  ],
  "declaration_order": [
    "fidl.test.handles/obj_type",
    "fidl.test.handles/handle",
    "fidl.test.handles/aliased_subtype_handle",
    "fidl.test.handles/aliased_rights_handle",
    "fidl.test.handles/aliased_plain_handle",
    "fidl.test.handles/Handles",
    "fidl.test.handles/SomeProtocol"
  ],
  "declarations": {
    "fidl.test.handles/obj_type": "enum",
    "fidl.test.handles/handle": "experimental_resource",
    "fidl.test.handles/SomeProtocol": "interface",
    "fidl.test.handles/Handles": "struct",
    "fidl.test.handles/aliased_plain_handle": "type_alias",
    "fidl.test.handles/aliased_subtype_handle": "type_alias",
    "fidl.test.handles/aliased_rights_handle": "type_alias"
  }
}
